### ISSUE TITLE
DS GET Command Overloaded

### ISSUE THREAD

**Issue by tobiasmo1** [edgexfoundry/device-sdk-go#240]
A device service implemented with the SDK currently handles GET commands by:
a) responding to caller with http **body** that contains the event/reading(s) emitted by the queried device endpoint
-and-
b) issues a corresponding **event** to core-data

In both cases the response body and event are encoded as application/json or application/cbor based on the data type of readings within the event content. Specifically if one or more readings are binary it implicitly encodes as application/cbor, otherwise as application/json.

To reduce overhead and optimize throughput, proposal is to introduce a new parameter (or other configurable control) to honor system/caller intention regarding how and/or whether a DS command will perform these two activities.

**Comment by SteveOss** [comment]
I think we want 3 possible behaviours here:

1 - Return event as a response, do not publish event to core data
2 - Do not return event as a response, publish event to core data
3 - Return event as a response, publish event to core data

Don't think not returning or publishing the event (possible option 4) makes any sense ?

**Comment by iain-anderson** [comment]
```
/device/<id>/<command>?publish=yes            <-- default, publish and return
/device/<id>/<command>?publish=no             <-- return event, do not publish
/device/<id>/<command>?publish=only           <-- publish event, do not return
```

**Comment by tobiasmo1** [comment]
Should publish=only be the default to serve primary workflow? It otherwise produces two operating paths.

```
/device/<id>/<command>                 <-- default, publish event, do not return
/device/<id>/<command>?response=yes    <-- publish event and return
/device/<id>/<command>?response=only   <-- return event, do not publish

```

**Comment by tobiasmo1** [comment]
My thinking here is to have default be prudent with memory usage at the edge; which could be exacerbated by embedding objects in value or due to potential size of binary payload.

**Comment by SteveOss** [comment]
@tobiasmo1 Agree that default should definitely not be "yes". Whether default is to return or publish
event I guess depends on what we think the typical usage is. If being called from external application
"only" makers sense, but not the case if being driven from scheduler.

**Comment by cloudxxx8** [comment]
if it is driven by AutoEvent, the behavior is only sending to Core Data without response event
AutoEvent calls handler directly, not controller

**Comment by tobiasmo1** [comment]
Lets enumerate potential callers to identify intended usage patterns.

DS command endpoint(s) get invoked by external apps and Support Scheduler. Also as result of condition match in Rules Engine? Other contexts?

**Comment by iain-anderson** [comment]
Another possibility - if the Readings cache is implemented, we may wish to give the option of sending to core-data only if there are changes.

**Comment by cloudxxx8** [comment]
@iain-anderson have you implemented this in C SDK?  Should we put into Hanoi scope?

**Comment by iain-anderson** [comment]
No, I haven't. It's a good feature, let's scope it for Hanoi

**Comment by cloudxxx8** [comment]
related to https://github.com/edgexfoundry/device-sdk-go/issues/543

**Comment by iain-anderson** [comment]
Note - this applies only for `api/v2/device` endpoints

**Comment by hahattan** [comment]
this is addressed in v2 command API (#619)
